### PR TITLE
Youtube caption corrections

### PR DESCRIPTION
**Pull request by 2dot71mily**
This PR adds a new dataset of YouTube captions, error and corrections. This dataset was created in just the last week, as inspired by this sprint!

### CODE REVIEW

**Comment by 2dot71mily** [comment]
Sorry about forgetting flake8.
Rather than use up the circleci resources on a new push with only formatting changes, I will wait to push until the results from all tests finish and/or any feedback comes in... probably tomorrow for me.

File: datasets/youtube_caption_corrections/README.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+size_categories:
+- n<1K
+source_datasets:
+- original
+task_categories:
+- other
+- sequence-modeling
+task_ids:
+- other-other-token-classification-of-text-errors
+- slot-filling
```
Comment (lhoestq) [inline datasets/youtube_caption_corrections/README.md:21]:
```suggestion
---
annotations_creators:
- expert-generated
- machine-generated
language_creators:
- machine-generated
languages:
- en
licenses:
- mit
multilinguality:
- monolingual
size_categories:
- n<1K
source_datasets:
- original
task_categories:
- other
- sequence-modeling
task_ids:
- other-other-token-classification-of-text-errors
- slot-filling
---
```

File: datasets/youtube_caption_corrections/youtube_caption_corrections.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+import datasets
+
+
+_CITATION = """\
+url={https://github.com/2dot71mily/youtube_captions_corrections}
+},
+year={2020}
+}
+"""
```
Comment (lhoestq) [inline datasets/youtube_caption_corrections/youtube_caption_corrections.py:30]: If there's no citation you can leave an empty string here

File: datasets/youtube_caption_corrections/youtube_caption_corrections.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    "TED",
+    "Weights_&_Biases",
+    "nature_video",
+    "Alfredo_Canziani",
+    "DeepMind",
+    "Pieter_Abbeel",
+    "Veritasium",
+    "minutephysics",
+    "stanfordonline",
+]
```
Comment (lhoestq) [inline datasets/youtube_caption_corrections/youtube_caption_corrections.py:57]: good content here :)

File: datasets/youtube_caption_corrections/youtube_caption_corrections.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+_URLS = [
+    f"https://raw.githubusercontent.com/2dot71mily/youtube_captions_corrections/main/data/transcripts/en/postproc_transcripts/{channel_name}.json"
+    for channel_name in _CHANNEL_NAMES
+]
+
+
+class YouTubeCaptionCorrections(datasets.GeneratorBasedBuilder):
```
Comment (lhoestq) [inline datasets/youtube_caption_corrections/youtube_caption_corrections.py:64]:
```suggestion
class YoutubeCaptionCorrections(datasets.GeneratorBasedBuilder):
```
this must be the camel case version of the dataset script name `youtube_caption_corrections`

You'll need to regenerate the dataset_infos.json file after this change

**Review by lhoestq — COMMENTED** [review]
Nice thanks !

I left a few comments.

Also it looks like the dummy_data.zip file is quite big (500KB).
Could you try to reduce its size please ? Ideally it should only be a few KB

To do so feel free to take a look inside the zip file and remove all the unnecessary files and chunks of text just to keep a few examples.

File: datasets/youtube_caption_corrections/youtube_caption_corrections.py (hunk 3 above)
Comment (2dot71mily) [reply datasets/youtube_caption_corrections/youtube_caption_corrections.py:57]: I had huggingface on the list, but the youtube channel doesn't seem to have any auto-gen, manually-corrected caption pairs, yet! However, I plan to keep occasionally rerunning my scripts to update the content per channel, and add more channels, too.

For this reason, I was originally hoping I could avoid hardcoding these channel names, and instead grab all jsons in the github repo dir.

I see you guys use the requests library to download the URLS, but I couldn't find a util in the code that would grab all links from a webpage request. Is that already out there, or would it be of interest, or not appropriate for this library?

**Comment by 2dot71mily** [comment]
So... my normal work is with mercurial and seem to have clearly forked this up using git... :(

What I did is after calling:
```
git fetch upstream
git rebase upstream/master
```

I then I attempt to pull in my most recent changes UI commit changes based on @lhoestq's feedback with:
```
git pull
```  
... which I now suspect undid the above fetch and rebase. Will look into fixing later today when I have more time. Sorry!

**Comment by 2dot71mily** [comment]
My dummy data seems quite large as a single row is composed of tokens/labels for an entire youtube video, with at least one row required for each file, which in this case 1 file per 13 youtube channels.

To make it smaller I passed `--n_lines 1` to reduce about 5x.

I then manually reduced size of the particularly long youtube lectures to get the size to about 30KB. However, after recompressing into a zip, and running dummy data test I got the following error:
`FAILED tests/test_dataset_common.py::LocalDatasetTest::test_load_dataset_all_configs_youtube_caption_corrections - OSError: Cannot find data file. `, despite file being there, which I haven't had a chance yet to debug.

**Comment by 2dot71mily** [comment]
I wrote a small script to generate a smaller json file for the dummy_data, with the hope that I could resolve the pytest error noted above (in case related to a manual typo I could have introduce), however the test contains to fail locally... here's to hoping it can pass on remote!

**Comment by 2dot71mily** [comment]
Sorry for delayed comments here. Last commit made two changes:
- Increased the valency of the labels from just True/False to more categories to describe the various types of diffs encountered. This required some rewrite of the README
- Reduced the number of remote files to be downloaded from 13 to 4, by combining all 13 of the channel-specific files together, and the splitting them up in a way to meet Github file size requirements. This also reduces size of the dummy-data.

File: datasets/youtube_caption_corrections/youtube_caption_corrections.py (hunk 3 above)
Comment (lhoestq) [reply datasets/youtube_caption_corrections/youtube_caption_corrections.py]: What we use is the `dl_manager` download methods. 
Basically what you can do is have one json in your repo with a list of all available channels.
Then in the script you can first download this json to get the list of channels, and them iterate over the channel names to download their json files.

Also you mentioned that you plan to re-run you script and add more channels.
That's a wonderful idea :) 
One suggestions I may add is have some some of versioning on your repo that we can use in the dataset script to specify which version we want to load (by date or using a specific github release tag for example). You want organize your versioning as you want. Either having one folder per date in the repo, or doing github releases for example.
What do you think ?

File: datasets/youtube_caption_corrections/README.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+[Needs More Information]
+
+### Discussion of Biases
+
+[Needs More Information]
+
+### Other Known Limitations
+
+[Needs More Information]
```
Comment (lhoestq) [inline datasets/youtube_caption_corrections/README.md:155]:
```suggestion
[More Information Needed]

### Discussion of Biases

[More Information Needed]

### Other Known Limitations

[More Information Needed]
```
we use [More Information Needed] in general

**Review by lhoestq — COMMENTED** [review]
This is really cool thanks !
Looks all good to me

To make the current code work even if you change the organisation of your repo, could you replace the URL with one that contains a commit sha ?

For example use `https://raw.githubusercontent.com/2dot71mily/youtube_captions_corrections/c85e2d7460b322e4a6062f01da5d1f4c06d05b39/data/transcripts/en/split/youtube_caption_corrections_0.json`

You'll need to regenerate the dataset_infos.json file after this change

File: datasets/youtube_caption_corrections/README.md (hunk 5 above)
Comment (2dot71mily) [reply datasets/youtube_caption_corrections/README.md:155]: Thanks, will do

**Comment by 2dot71mily** [comment]
@lhoestq, thank you for the great feedback, especially given how busy you guys are now! 

I checked out GitHub release tags and looks cool. I have added the version tag to the url, instead of the commit sha as originally suggested, with the hope that it serves the same purpose of pinning the content to this url. Please let me know if I have misunderstood.

In regard to dynamically changing the number of files downloaded by first downloading a JSON listing the files, I love that idea. But I am a little confused, as I was thinking that any changes to the dataset itself would require a new PR with an updated `dataset_infos.json`, e.g. `num_examples` would increase. 

If the purpose of this is not to permit dynamic (without a PR needed) growth of the number of files, but instead to provide stability to the consumers of the dataset, maybe I continued use the release tags, maintaining access to old releases could serve this purpose? I am still learning about these release tags...

**Comment by lhoestq** [comment]
For dynamic datasets, i.e. datasets that evolve over time, we support custom configurations: they are configurations that are not part of the BUILDER_CONFIGS or in the dataset_infos.json

For example for wikipedia, you can use the latest wiki dump by specifying `date=` inside `load_dataset()`. A configuration is created on the fly for this date and is used to build the dataset using the latest data.

Therefore we don't need to have PRs to update the script for each wikipedia release.

One downside though is that we don't have metadata in advance such as the size of the dataset.

I think this could be a nice addition for the youtube caption dataset in the future to be have a system of releases and be able to load the latest one easily. What do you think ?

**Review by lhoestq — APPROVED** [review]
Looks all good to me :) 

Maybe let's merge this one for now and see how we can support the future release later ?

**Comment by 2dot71mily** [comment]
> For dynamic datasets, i.e. datasets that evolve over time, we support custom configurations: they are configurations that are not part of the BUILDER_CONFIGS or in the dataset_infos.json
> 
 
> I think this could be a nice addition for the youtube caption dataset in the future to be have a system of releases and be able to load the version we want easily. What do you think ?

Thank you for the suggestion! This sounds great! I will take a look at the some datasets that do this, and would love to give it a try in the future, if I continue to grow the captions dataset in a meaningful way. 

Appreciate all the help on this. It has been a really great experience for me. :)

**Comment by 2dot71mily** [comment]
Excited to merge! And sorry to be such a github n00b, but from what I've quickly read, I don't 'Close pull request', but rather the next steps are action taken on your end... Please let me know if there is some action to be taken at my end first. :/

**Comment by lhoestq** [comment]
Alright merging this one then :)

**Pull request merged**